packages/coinjoin/src/client/round/inputRegistration.ts

```diff
@@ -0,0 +1,185 @@
+            zeroVsizeCredentials,
+            {
+                signal,
+                baseUrl: coordinatorUrl,
+                identity: input.outpoint,
+                delay,
+            },
+        )
+        .catch(error => {
+            options.log(
+                `Registration <${input.outpoint}> to <${round.id}> failed: ${error.message}`,
+            );
+            // catch specific error
+            if (error.message === coordinator.WabiSabiProtocolErrorCode.WrongPhase) {
+                // abort remaining delayed candidates to register (if exists) registration is not going to happen for them anyway
+                signal.dispatchEvent(new Event('abort'));
+            }
+            throw error;
+        });
+
+    // store RegistrationData
+    input.setRegistrationData(registrationData);
+    // and put input to prison
+    prison.detain(input.outpoint, {
+        roundId: round.id,
+        reason: coordinator.WabiSabiProtocolErrorCode.AliceAlreadyRegistered,
+    });
+
+    // NOTE: RegistrationData processing on middleware is intentionally not using abort signal
+    // should not be aborted if round phase was immediately changed after registration (triggered by Status change)
+    try {
+        // process RegistrationData received from coordinator
+        // get Credentials and use them in middleware.getRealCredentials
+        const amountCredentials = await middleware.getCredentials(
+            round.amountCredentialIssuerParameters,
+            registrationData.amountCredentials,
+            zeroAmountCredentials.credentialsResponseValidation,
+            { baseUrl: middlewareUrl }, // NOTE: without abort signal (should not be aborted)
+        );
+        const vsizeCredentials = await middleware.getCredentials(
+            round.vsizeCredentialIssuerParameters,
+            registrationData.vsizeCredentials,
+            zeroVsizeCredentials.credentialsResponseValidation,
+            { baseUrl: middlewareUrl }, // NOTE: without abort signal (should not be aborted)
+        );
+
+        // Calculate mining and coordinator fee
+        // coordinator fee is 0 if input is remixed or amount is lower than plebsDontPayThreshold value
+        const { roundParameters } = round;
+        const coordinatorFee =
+            input.amount > roundParameters.coordinationFeeRate.plebsDontPayThreshold &&
+            !registrationData.isPayingZeroCoordinationFee
+                ? Math.floor(roundParameters.coordinationFeeRate.rate * input.amount)
+                : 0;
+
+        const miningFee = Math.floor((input.inputSize * roundParameters.miningFeeRate) / 1000);
+        const amount = input.amount - coordinatorFee - miningFee;
+        const vsize = roundParameters.maxVsizeAllocationPerAlice - input.inputSize;
+
+        // use Credentials to get RealCredentials. use them in outputConfirmation
+        const realAmountCredentials = await middleware.getRealCredentials(
+            [amount, 0],
```

Comment:
what the `0` as second amount means here?

Reply:
it's a nature of coordinator format.
RealCredentials[] are always "a pair" its useful when you want to join two Credentials (see outputDemposition) but in this case you want for fill (or in other words you have data) only for 1 Credential